tests/core/middleware/test_name_to_address_middleware.py

```diff
@@ -0,0 +1,33 @@
+import pytest
+
+from ens import ENS
+from web3 import Web3
+from web3.exceptions import (
+    InvalidAddress,
+)
+from web3.middleware import (  # noqa: F401
+    name_to_address_middleware,
+)
+from web3.providers import (
+    HTTPProvider,
+)
+
+
+def w3_name_resolver(url):
+    w3 = Web3(providers=[HTTPProvider(url)], middlewares=[])
+    w3.ens = ENS.fromWeb3(w3)
+    w3.middleware_stack.add(name_to_address_middleware(w3))
+    return w3
+
+
+def test_pass_name_resolver_mainnet():
+    url = "https://mainnet.infura.io/"
+    w3 = w3_name_resolver(url)
+    w3.eth.getBalance("ethereum.eth")
```

Comment:
This looks like it's actually hitting Infura for the test, right? Tests should be able to run while disconnected from the Internet.

One option is to use `construct_fixture_middleware` to return the network id (probably injected at layer 0).

Reply:
Pushed the test offline but the BaseProvider was missing several methods which had to be implemented eg `eth_getBlockByNumber`, `eth_call` and `eth_getBalance`.
Tried to enable these methods using `construct_fixture_middleware` at layer 0 but the command `w3.eth.getBalance(name)` would call `eth_getBalance` --> `eth_call` --> `eth_getBlockByNumber` (which would again throw method not implemented error).

Maybe after 2-3 recusrive calls the method `construct_fixture_middleware` fails.

Have bypassed this issue by letting pytest catch the `NotImplementedError` do tell me if this is not an acceptable solution.

Reply:
@carver drive-by-idea.

- `fixture_recorder_middleware`

Lets you connect whatever provider you like, make some calls, and then export the request/response data in a format that could then be passed into `construct_fixture_middleware` to emulate the request/responses.

Reply:
Ooh, neat idea.